implicit/als.py

```diff
@@ -105,12 +105,36 @@ def recommend(self, userid, user_items, N=10, filter_items=None, recalculate_use
     def _user_factor(self, userid, user_items, recalculate_user=False):
         if not recalculate_user:
             return self.user_factors[userid]
-        Y = self.item_factors
-        if self._YtY is None:
-            self._YtY = Y.T.dot(Y)
-        return user_factor(Y, self._YtY, user_items, userid,
+        return user_factor(self.item_factors, self.YtY, user_items, userid,
                            self.regularization, self.factors)
 
+    def explain(self, userid, user_items, itemid, user_weights=None, N=10):
+        """ Returns the predicted rating for an user x item pair,
+            the explanation (the contribution from the top N items the user liked),
+            and a user latent factor weight that can be cached if you want to
+            get more than one explanation for the same user.
+        """
+        if user_weights is None:
+            A, _ = user_linear_equation(self.item_factors, self.YtY,
+                                        user_items, userid,
+                                        self.regularization, self.factors)
+            user_weights = np.linalg.inv(A)
```

Comment:
I'm not against using LU factorization here - but I think cholesky factorization will work, and will be much faster.

So 'A' is always at least a [positive semi-definite matrix](https://en.wikipedia.org/wiki/Positive-definite_matrix#Positive-semidefinite) (YtY squares the Y matrix making A >= 0 - and as long as confidence and regularization are positive this doesn't change).  I also believe that adding in a non-zero positive regularization factor makes this matrix positive definite (instead of just semi-definite) by preventing ```zT A z``` from being 0 for any row/column 'z' of A.

[This paper](https://pdfs.semanticscholar.org/bfdf/7af6cf7fd7bb5e6b6db5bbd91be11597eaf0.pdf) mentions using the cholesky decomposition to solve A, and I'm using a cholesky solver in the cython version here: https://github.com/benfred/implicit/blob/master/implicit/_als.pyx#L103  (and yeah, I know I have some code to fallback to LU  there if it fails (which in theory should only be for 0 regularization) - but it doesn't actually work,  and probably should be removed since I haven't seen it get called successfully yet=).